server/src/gnome/prompter.rs

```diff
@@ -0,0 +1,407 @@
+                )
+                .await?;
+            }
+            PromptRole::Unlock => {
+                self.prompt(
+                    false,
+                    reply,
+                    exchange,
+                    connection.clone(),
+                    prompt.path().clone(),
+                    prompt.objects().clone(),
+                )
+                .await?;
+            }
+            PromptRole::CreateCollection => todo!(),
+        };
+
+        Ok(())
+    }
+
+    pub async fn prompt_done(
+        &self,
+        #[zbus(object_server)] object_server: &zbus::ObjectServer,
+    ) -> Result<(), ServiceError> {
+        if let Some(prompt) = self.service.prompt().await {
+            object_server.remove::<Prompt, _>(prompt.path()).await?;
+            self.service.remove_prompt().await;
+        }
+        object_server.remove::<Self, _>(&self.path).await?;
+
+        Ok(())
+    }
+}
+
+impl PrompterCallback {
+    pub async fn new(service: Service) -> Self {
+        let index = service.prompt_index().await;
+        Self {
+            path: OwnedObjectPath::try_from(format!("/org/gnome/keyring/Prompt/p{index}")).unwrap(),
+            service,
+        }
+    }
+
+    pub fn path(&self) -> &OwnedObjectPath {
+        &self.path
+    }
+
+    pub async fn object_label(&self, objects: &Vec<OwnedObjectPath>) -> Option<String> {
+        for object in objects {
+            if let Some(collection) = self.service.collection_from_path(object).await {
+                return Some(collection.label().await);
+            }
+        }
+        None
+    }
+
+    pub async fn perform_prompt(
+        connection: zbus::Connection,
+        path: OwnedObjectPath,
+        prompt_type: PromptType,
+        properties: Properties,
+        exchange: String,
+    ) -> Result<(), ServiceError> {
+        let prompter = PrompterProxy::new(&connection).await?;
+        prompter
+            .perform_prompt(path, prompt_type, properties, &exchange)
+            .await?;
+
+        Ok(())
+    }
+
+    pub async fn stop_prompting(
+        connection: zbus::Connection,
+        path: OwnedObjectPath,
+    ) -> Result<(), ServiceError> {
+        let prompter = PrompterProxy::new(&connection).await?;
+        prompter.stop_prompting(path).await?;
+
+        Ok(())
+    }
+
+    pub async fn prompt_completed(
+        signal_emitter: zbus::object_server::SignalEmitter<'_>,
+        dismissed: bool,
+        result: OwnedValue,
+    ) -> Result<(), ServiceError> {
+        Prompt::completed(&signal_emitter, dismissed, result).await?;
+        tracing::debug!("Prompt completed.");
+
+        Ok(())
+    }
+
+    pub async fn prompt(
+        &self,
+        locked: bool,
+        reply: Optional<Reply>,
+        exchange: &str,
+        connection: zbus::Connection,
+        prompt_path: OwnedObjectPath,
+        objects: Vec<OwnedObjectPath>,
+    ) -> Result<(), ServiceError> {
+        if let Some(reply) = reply.as_ref() {
```

Comment:
why this and the match at the same time?

Reply:
Reply (has just two variants now) is wrapped around `zvariant::Optional`. So, here if there is a value we'll match it, otherwise `else` block will handle the NoneValue which is `''`.

Reply:
Optional implements deref to Option so you should use that.

Reply:
Done, thanks!